rhapsody/heatmap.py

```diff
@@ -0,0 +1,157 @@
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+
+_heatmap_choices = {
+    'metric': {'braycurtis', 'canberra', 'chebyshev', 'cityblock',
+               'correlation', 'cosine', 'dice', 'euclidean', 'hamming',
+               'jaccard', 'kulsinski', 'mahalanobis', 'matching', 'minkowski',
+               'rogerstanimoto', 'russellrao', 'seuclidean', 'sokalmichener',
+               'sokalsneath', 'sqeuclidean', 'yule'},
+    'method': {'single', 'complete', 'average', 'weighted', 'centroid',
+               'median', 'ward'}}
+
+_cmaps = {
+    'heatmap': [
+        'PiYG', 'PRGn', 'BrBG', 'PuOr', 'RdGy', 'RdBu',
+        'RdYlBu', 'RdYlGn', 'Spectral', 'coolwarm', 'bwr', 'seismic',
+        'viridis', 'plasma', 'inferno', 'magma', 'cividis'],
+    'margins': [
+        'cubehelix', 'Pastel1', 'Pastel2', 'Paired',
+        'Accent', 'Dark2', 'Set1', 'Set2', 'Set3',
+        'tab10', 'tab20', 'tab20b', 'tab20c',
+        'viridis', 'plasma', 'inferno', 'magma', 'cividis']}
+
+
+def ranks_heatmap(ranks, microbe_metadata=None, metabolite_metadata=None,
+                  method='average', metric='euclidean',
+                  color_palette='seismic', margin_palette='cubehelix',
+                  x_labels=False, y_labels=False, level=3, threshold=3):
+    '''
+    Generate clustermap of microbe X metabolite conditional probabilities.
+
+    ranks: pd.DataFrame of conditional probabilities.
+        Microbes (rows) X metabolites (columns).
+    microbe_metadata: pd.Series of microbe metadata for annotating plots
+    metabolite_metadata: pd.Series of metabolite metadata for annotating plots
+    method: str
+        Hierarchical clustering method used in clustermap.
+    metric: str
+        Hierarchical clustering distance metric used in clustermap.
+    color_palette: str
+        Color palette for clustermap.
+    margin_palette: str
+        Name of color palette to use for annotating metadata
+        along margin(s) of clustermap.
+    x_labels: bool
+        Plot x-axis (metabolite) labels?
+    y_labels: bool
+        Plot y-axis (microbe) labels?
+    level: int
+        taxonomic level for annotating clustermap. Set to -1 if not parsing
+        semicolon-delimited taxonomies or wish to print entire annotation.
+    threshold: int
+        Minimum absolute value of conditional probabilities to plot.
+        metabolites/taxa that do not have at least one CP above the
+        threshold will be removed.
+
+    Returns clustermap
+    '''
+    # filter microbes
+    abs_vals = ranks.abs()
+    ranks = ranks[abs_vals.max(axis=1) > threshold]
```

Comment:
This is one of the more challenging parts for the UI.

First, definitely don't want to take the max of the absolutes - we don't care about the strongly negative conditional probabilities (those are typically not informative). 

Since the user is already passing in metadata, maybe we could just have the ranks just match the metadata and only consider the intersection?  The filtering criteria can get very complicated otherwise (i.e. for instance, the threshold is expected not to be reliable compared to the ranks.  Also the negative conditional log probabilities won't be informative, so we won't want to take the absolute values of that).